testtools/files.go

```diff
@@ -129,3 +139,232 @@ func CheckFiles(t *testing.T, dir string, files []FileSpec) {
+	// Default: ".in", so input BUILD files should be named BUILD.in.
+	BuildInSuffix string
+
+	// The suffix for all test output build files. Includes the ".".
+	// Default: ".out", so out BUILD files should be named BUILD.out.
+	BuildOutSuffix string
+}
+
+func NewTestGazelleGenerationArgs() *TestGazelleGenerationArgs {
+	return &TestGazelleGenerationArgs{
+		BuildInSuffix:  ".in",
+		BuildOutSuffix: ".out",
+	}
+}
+
+// TestGazelleGenerationOnPath runs a full gazelle binary on a testdata directory.
+// With a test data directory of the form:
+//└── <testDataPath>
+//    └── some_test
+//        ├── WORKSPACE
+//        ├── README.md --> README describing what the test does.
+//        ├── test.yaml --> YAML file for test configuration.
+//        └── app
+//            └── sourceFile.foo
+//            └── BUILD.in --> BUILD file prior to running gazelle.
+//            └── BUILD.out --> BUILD file expected after running gazelle.
+func TestGazelleGenerationOnPath(t *testing.T, args *TestGazelleGenerationArgs, files []bazel.RunfileEntry) {
+	t.Run(args.Name, func(t *testing.T) {
+		var inputs []FileSpec
+		var goldens []FileSpec
+
+		var config *testYAML
+		for _, f := range files {
+			path := f.Path
+			trim := filepath.Join(args.TestDataPath, args.Name) + "/"
+			shortPath := strings.TrimPrefix(f.ShortPath, trim)
+
+			info, err := os.Stat(path)
+			if err != nil {
+				t.Fatalf("os.Stat(%q) error: %v", path, err)
+			}
+
+			if info.IsDir() {
+				continue
+			}
+
+			content, err := ioutil.ReadFile(path)
+			if err != nil {
+				t.Errorf("ioutil.ReadFile(%q) error: %v", path, err)
+			}
+
+			if filepath.Base(shortPath) == "test.yaml" {
+				if config != nil {
+					t.Fatal("only 1 test.yaml is supported")
+				}
+				config = new(testYAML)
+				if err := yaml.Unmarshal(content, config); err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			if strings.HasSuffix(shortPath, args.BuildInSuffix) {
+				inputs = append(inputs, FileSpec{
+					Path:    filepath.Join(args.Name, strings.TrimSuffix(shortPath, args.BuildInSuffix)+".bazel"),
+					Content: string(content),
+				})
+			} else if strings.HasSuffix(shortPath, args.BuildOutSuffix) {
+				goldens = append(goldens, FileSpec{
+					Path:    filepath.Join(args.Name, strings.TrimSuffix(shortPath, args.BuildOutSuffix)+".bazel"),
+					Content: string(content),
+				})
+			} else {
+				inputs = append(inputs, FileSpec{
+					Path:    filepath.Join(args.Name, shortPath),
+					Content: string(content),
+				})
+				goldens = append(goldens, FileSpec{
+					Path:    filepath.Join(args.Name, shortPath),
+					Content: string(content),
+				})
+			}
+		}
+
+		testdataDir, cleanup := CreateFiles(t, inputs)
+		defer cleanup()
+		defer func() {
+			if t.Failed() {
+				shouldUpdate := os.Getenv("UPDATE_SNAPSHOTS") != ""
+				buildWorkspaceDirectory := os.Getenv("BUILD_WORKSPACE_DIRECTORY")
+				if buildWorkspaceDirectory == "" {
+					// Default to ~/<workspace>.
```

Comment:
If `BUILD_WORKSPACE_DIRECTORY` is empty, I think we should just `t.Fatal` out instead of trying to guess where things should be written. We are probably going to guess wrong and we should reinstruct the caller to use `bazel run`

Reply:
Yep, totally agree.